packages/react-router/tests/utils.test.ts

```diff
@@ -0,0 +1,38 @@
+import { describe, it, expect } from 'vitest'
+import { replaceEqualDeep } from '../src/utils';
+
+describe('replaceEqualDeep', () => {
+  it('should return the same object if the input objects are equal', () => {
+    const obj = { a: 1, b: 2 };
+    const result = replaceEqualDeep(obj, obj);
+    expect(result).toBe(obj);
+  });
+
+  it('should return a new object with replaced values if the input objects are not equal', () => {
+    const obj1 = { a: 1, b: 2 };
+    const obj2 = { a: 1, b: 3 };
+    const result = replaceEqualDeep(obj1, obj2);
+    expect(result).toStrictEqual(obj2);
+  });
+
+  it('should handle arrays correctly', () => {
+    const arr1 = [1, 2, 3];
+    const arr2 = [1, 2, 4];
+    const result = replaceEqualDeep(arr1, arr2);
+    expect(result).toStrictEqual(arr2);
+  });
+
+  it('should handle nested objects correctly', () => {
+    const obj1 = { a: 1, b: { c: 2 } };
+    const obj2 = { a: 1, b: { c: 3 } };
+    const result = replaceEqualDeep(obj1, obj2);
+    expect(result).toStrictEqual(obj2);
+  });
+
+  it('should properly handle non-existent keys', () => {
+    const obj1 = { a: 2, c: 123 }
+    const obj2 = { a: 2, b: undefined };
+    const result = replaceEqualDeep(obj1, obj2);
+    expect(result).toStrictEqual(obj2);
```

Comment:
@tannerlinsley  is this expectation correct?

the function's comment says:

> This function returns `a` if `b` is deeply equal.
> If not, it will replace any deeply equal children of `b` with those of `a`.

What about children that exist in `b` but not in `a`?

Reply:
If children exist in B, then a new object will be returned with anything else that hasn't changed from A, but with the new stuff from B:

So it's possible than an object would be a mix:

{
  foo: A,
  bar: B,
  sub: {
    foo: B
  }
}